p2p/background/router_test.go

```diff
@@ -0,0 +1,373 @@
+
+	if libp2pMockNet == nil {
+		libp2pMockNet = mocknet.New()
+	}
+
+	host := newMockNetHostFromPeer(t, libp2pMockNet, privKey, selfPeer)
+	t.Cleanup(func() {
+		err := host.Close()
+		require.NoError(t, err)
+	})
+
+	return newRouterWithSelfPeerAndHost(t, selfPeer, host)
+}
+
+func newRouterWithSelfPeerAndHost(t *testing.T, selfPeer typesP2P.Peer, host libp2pHost.Host) *backgroundRouter {
+	t.Helper()
+
+	ctrl := gomock.NewController(t)
+	runtimeMgrMock := mockModules.NewMockRuntimeMgr(ctrl)
+	runtimeMgrMock.EXPECT().GetConfig().Return(&configs.Config{
+		P2P: &configs.P2PConfig{
+			IsClientOnly: false,
+		},
+	})
+
+	consensusMock := mockModules.NewMockConsensusModule(ctrl)
+	consensusMock.EXPECT().CurrentHeight().Return(uint64(1)).AnyTimes()
+
+	pstore := make(typesP2P.PeerAddrMap)
+	pstoreProviderMock := mock_types.NewMockPeerstoreProvider(ctrl)
+	pstoreProviderMock.EXPECT().GetStakedPeerstoreAtHeight(gomock.Any()).Return(pstore, nil).AnyTimes()
+
+	busMock := mockModules.NewMockBus(ctrl)
+	busMock.EXPECT().GetConsensusModule().Return(consensusMock).AnyTimes()
+	busMock.EXPECT().GetRuntimeMgr().Return(runtimeMgrMock).AnyTimes()
+
+	err := pstore.AddPeer(selfPeer)
+	require.NoError(t, err)
+
+	router, err := NewBackgroundRouter(busMock, &utils.RouterConfig{
+		Addr:                  selfPeer.GetAddress(),
+		PeerstoreProvider:     pstoreProviderMock,
+		CurrentHeightProvider: consensusMock,
+		Host:                  host,
+	})
+	require.NoError(t, err)
+
+	libp2pNet, ok := router.(*backgroundRouter)
+	require.Truef(t, ok, "unexpected router type: %T", router)
+
+	return libp2pNet
+}
+
+// TECHDEBT: move & de-dup
+func newTestPeer(t *testing.T) (cryptoPocket.PrivateKey, *typesP2P.NetworkPeer) {
+	t.Helper()
+
+	privKey, err := cryptoPocket.GeneratePrivateKey()
+	require.NoError(t, err)
+
+	return privKey, &typesP2P.NetworkPeer{
+		PublicKey:  privKey.PublicKey(),
+		Address:    privKey.Address(),
+		ServiceURL: testLocalServiceURL,
+	}
+}
+
+func newMockNetHostFromPeer(
+	t *testing.T,
+	mockNet mocknet.Mocknet,
+	privKey cryptoPocket.PrivateKey,
+	peer *typesP2P.NetworkPeer,
+) libp2pHost.Host {
+	t.Helper()
+
+	libp2pPrivKey, err := libp2pCrypto.UnmarshalEd25519PrivateKey(privKey.Bytes())
+	require.NoError(t, err)
+
+	libp2pMultiAddr, err := utils.Libp2pMultiaddrFromServiceURL(peer.ServiceURL)
+	require.NoError(t, err)
+
+	host, err := mockNet.AddPeer(libp2pPrivKey, libp2pMultiAddr)
+	require.NoError(t, err)
+
+	return host
+}
+
+func newTestHost(t *testing.T, mockNet mocknet.Mocknet, privKey cryptoPocket.PrivateKey) libp2pHost.Host {
+	t.Helper()
+
+	// listen on random port on loopback interface
+	peer := &typesP2P.NetworkPeer{
+		PublicKey:  privKey.PublicKey(),
+		Address:    privKey.Address(),
+		ServiceURL: testLocalServiceURL,
+	}
+
+	// construct mock host
+	return newMockNetHostFromPeer(t, mockNet, privKey, peer)
+}
+
+func readSubscription(
+	t *testing.T,
+	ctx context.Context,
+	wg *sync.WaitGroup,
```

Comment:
Can you reflect the same name of the wg as what you passed in. It'll help with the flow of the code to the reader.